### PR TITLE
remove memory allocation in WatcherImpl for Windows

### LINKED ISSUE #14307: Remove memory  allocation in Win32 WatcherImpl::directoryChangeCompletion
In `WatcherImpl::directoryChangeCompletion` we allocate a `Buffer::OwnedImpl` to call `ioHandle` `write`.

This should be changed to `writev` to prevent memory allocation or we should implement and use #14244

### PR DESCRIPTION
**Pull request by rmiller14**
Signed-off-by: Randy Miller <rmiller14@gmail.com>

Commit Message: Remove memory allocation in WatcherImpl::directoryChangeCompletion for Windows by replacing ioHandle write (which requires Buffer::OwnedImpl) with ioHandle writev.

Risk Level: Low
Testing: Ran the watcher_impl_test filesystem test 1000 times with no regressions.
Issues: Fixes #14307.

### CODE REVIEW

**Comment by rmiller14** [comment]
cc @envoyproxy/windows-dev

**Review by davinci26 — APPROVED** [review]
Thanks! LGTM

File: source/common/filesystem/win32/watcher_impl.cc [hunk 1]
```diff
@@ -194,7 +194,7 @@ void WatcherImpl::directoryChangeCompletion(DWORD err, DWORD num_bytes, LPOVERLA
       events |= Events::Modified;
     }
 
-    constexpr absl::string_view data{"a"};
+    char data[] = {'a'};
```
Comment (lizan) [inline source/common/filesystem/win32/watcher_impl.cc:197]: why did you need to change this? string_view can be used as well.

Comment (rmiller14) [reply source/common/filesystem/win32/watcher_impl.cc:197]: I changed it since keeping it would require a cast at line 208 to get rid of const, as the RawSlice constructor expects the first parameter to be void *, not const void *.  The char array is also smaller than string_view, though that hardly matters.

I guess there's no explicit need for this change though, so I'm happy to change it back to minimize number of lines churned.

Comment (lizan) [reply source/common/filesystem/win32/watcher_impl.cc:197]: > The char array is also smaller than string_view, though that hardly matters.

While char array is smaller than string_view, a `constexpr` implies `static` so there's no allocation in runtime. I prefer keep this constexpr.

Comment (rmiller14) [reply source/common/filesystem/win32/watcher_impl.cc:197]: Sure, I'll update.

**Comment by lizan** [comment]
can you check format?

**Comment by rmiller14** [comment]
> can you check format?

Fixed in latest commit.

**Review by lizan — APPROVED** [review]

**Comment by lizan** [comment]
can you merge main branch?

**Comment by rmiller14** [comment]
Done.

**Comment by rmiller14** [comment]
@lizan can this get another review?

**Review by mattklein123 — APPROVED** [review]
